plugins/inputs/huawei_routers_telemetry/huawei_routers_telemetry.go

```diff
@@ -0,0 +1,219 @@
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type setReadBufferer interface {
+	SetReadBuffer(bytes int) error
+}
+
+type streamSocketListener struct {
+	net.Listener
+	*HuaweiRoutersTelemetry
+
+	sockType string
+
+	connections    map[string]net.Conn
+	connectionsMtx sync.Mutex
+}
+
+type packetSocketListener struct {
+	net.PacketConn
+	*HuaweiRoutersTelemetry
+}
+
+/*
+  Telemetry Decoder.
+
+*/
+func HuaweiTelemetryDecoder(body []byte) (*metric.SeriesGrouper, error) {
+	msg := &telemetry.Telemetry{}
+	err := proto.Unmarshal(body[12:], msg)
+	if err != nil {
+		fmt.Println("Unable to decode incoming packet: ", err.Error())
+		return nil, err
+		//panic(err)
+	}
+	grouper := metric.NewSeriesGrouper()
+	for _, gpbkv := range msg.GetDataGpb().GetRow() {
+		dataTime := gpbkv.Timestamp
+		if dataTime == 0 {
+			dataTime = msg.MsgTimestamp
+		}
+		timestamp := time.Unix(int64(dataTime/1000), int64(dataTime%1000)*1000000)
+		sensorMsg := huawei_sensorPath.GetMessageType(msg.GetSensorPath())
+		err = proto.Unmarshal(gpbkv.Content, sensorMsg)
+		if err != nil {
+			fmt.Println("Sensor Error: ", err.Error())
+			return nil, err
+		}
+		fields, vals := huawei_sensorPath.SearchKey(gpbkv, msg.GetSensorPath())
+		tags := make(map[string]string, len(fields)+3)
+		tags["source"] = msg.GetNodeIdStr()
+		tags["subscription"] = msg.GetSubscriptionIdStr()
+		tags["path"] = msg.GetSensorPath()
+		// Search for Tags
+		for i := 0; i < len(fields); i++ {
+			tags = huawei_sensorPath.AppendTags(fields[i], vals[i], tags, msg.GetSensorPath())
+		}
+		// Create Metrics
+		for i := 0; i < len(fields); i++ {
+			huawei_sensorPath.CreateMetrics(grouper, tags, timestamp, msg.GetSensorPath(), fields[i], vals[i])
+		}
+	}
+	return grouper, nil
+}
+
+/*
+  Listen UDP packets and call the telemetryDecoder.
+*/
+func (h *packetSocketListener) listen() {
+	buf := make([]byte, 64*1024) // 64kb - maximum size of IP packet
+	for {
+		n, _, err := h.ReadFrom(buf)
+		if err != nil {
+			h.Log.Error("Unable to read buffer: %s", err.Error())
+			break
+		}
+
+		body, err := h.decoder.Decode(buf[:n])
+		if err != nil {
+			h.Log.Errorf("Unable to decode incoming packet: %s", err.Error())
+			continue
+		}
+		// Telemetry parsing over packet payload
+		grouper, err := HuaweiTelemetryDecoder(body)
+		if err != nil {
+			h.Log.Errorf("Unable to decode telemetry information: %s", err.Error())
+			break
+		}
+		for _, metric := range grouper.Metrics() {
+			h.AddMetric(metric)
+		}
+
+		if err != nil {
+			h.Log.Errorf("Unable to parse incoming packet: %s", err.Error())
+		}
+	}
+}
+
+type HuaweiRoutersTelemetry struct {
+	ServicePort     string        `toml:"service_port"`
+	ReadBufferSize  internal.Size `toml:"read_buffer_size"`
+	ContentEncoding string        `toml:"content_encoding"`
+	wg              sync.WaitGroup
+
+	Log telegraf.Logger `toml:"-"`
+
+	telegraf.Accumulator
+	io.Closer
+	decoder internal.ContentDecoder
```

Comment:
Please group by external/internal/anonymous:
```suggestion
	Log telegraf.Logger `toml:"-"`

	decoder internal.ContentDecoder
	wg         sync.WaitGroup

		telegraf.Accumulator
	io.Closer
```

Reply:
Oh and add a variable for `telegraf.Accumulator` such as `acc telegraf.Accumulator`.

Reply:
done